src/storage-controller/src/lib.rs

```diff
@@ -748,16 +749,58 @@ where
                             self.introspection_tokens.insert(id, scraper_token);
                         }
                         IntrospectionType::SourceStatusHistory => {
-                            self.partially_truncate_status_history(
-                                IntrospectionType::SourceStatusHistory,
+                            let last_status_per_id = self
+                                .partially_truncate_status_history(
+                                    IntrospectionType::SourceStatusHistory,
+                                )
+                                .await;
+
+                            let status_col = collection_status::MZ_SOURCE_STATUS_HISTORY_DESC
+                                .get_by_name(&ColumnName::from("status"))
+                                .expect("schema has not changed")
+                                .0;
+
+                            self.collection_status_manager.extend_previous_statuses(
+                                last_status_per_id.into_iter().flatten().map(|(id, row)| {
+                                    (
+                                        id,
+                                        Status::from_str(
+                                            row.iter()
+                                                .nth(status_col)
+                                                .expect("schema has not changed")
+                                                .unwrap_str(),
+                                        )
+                                        .expect("statuses must be uncorrupted"),
+                                    )
+                                }),
                             )
-                            .await;
                         }
                         IntrospectionType::SinkStatusHistory => {
```

Comment:
these can probably be merged into one branch right?

Reply:
You mean the branches off of `IntrospectionType::SourceStatusHistory` and `IntrospectionType::SinkStatusHistory`? Annoyingly not so easily because their `status_col`s come from different collections